### PR TITLE
Fixes #30715 - Add foreman JS tests github action

### PR DESCRIPTION
**Pull request by Ron-Lavi**
This GH action can run in parallel with travis until we decide to migrate for good

### CODE REVIEW

**Comment by Ron-Lavi** [comment]
@tbrisker @ezr-ondrej can you take a look?

File: webpack/assets/javascripts/react_app/redux/middlewares/index.js [hunk 1]
```diff
@@ -15,6 +15,10 @@ const logReduxToConsole = () => {
   return isProduction && isLogger;
 };
 
+/**
+ * The order of middlewares is important,
+ * the action will be passed between middlwares in the order they are written.
+ * read more about it in https://redux.js.org/api/applymiddleware */
```
Comment (ekohl) [inline webpack/assets/javascripts/react_app/redux/middlewares/index.js]: Is this intentional?

File: .github/workflows/js_tests.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
+name: JavaScript Testing
+on:
+  pull_request:
+    paths:
+      - 'webpack/**'
+      - 'package.json'
+      - '.github/workflows/js_tests.yml'
+
+jobs:
+  test_js:
```
Comment (ekohl) [inline .github/workflows/js_tests.yml]: I'd keep it short
```suggestion
  test:
```

File: .github/workflows/js_tests.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+name: JavaScript Testing
```
Comment (ekohl) [inline .github/workflows/js_tests.yml]: I'd keep it short
```suggestion
name: JavaScript
```

**Review by ekohl — COMMENTED** [review]
This makes sense, but I'd be in favor of dropping the Travis job in the same commit.

File: webpack/assets/javascripts/react_app/redux/middlewares/index.js (hunk 1 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/redux/middlewares/index.js]: Yeah I wanted to trigger travis to work so we could compare gh actions with it.
Plus a nice to have comment :)

**Comment by Ron-Lavi** [comment]
@ekohl so should we remove travis or let them live along each other for a few weeks or so?

**Comment by ekohl** [comment]
I'd drop Travis. I doubt anyone will compare the the two and in practice nobody will look at the actual results unless they fail.

**Comment by Ron-Lavi** [comment]
Goodbye Travis  :stuck_out_tongue: 
Ready for another review

File: .github/workflows/js_tests.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      - 'package.json'
+      - '.github/workflows/js_tests.yml'
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        node-version: [10, 12]
```
Comment (ekohl) [inline .github/workflows/js_tests.yml]: I always like the multi line notation more (as was done in Travis). When you add or remove a version, the diff is smaller and easier to read.

**Review by ekohl — DISMISSED** [review]
Small nit, but otherwise :+1:

File: .github/workflows/js_tests.yml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        node-version: [
+          10,
+          12,
+        ]
```
Comment (ekohl) [inline .github/workflows/js_tests.yml]: This is the YAML syntax for it, exactly the same as in `.travis.yml`.
```suggestion
        node-version:
          - 10
          - 12
```

**Review by ekohl — CHANGES_REQUESTED** [review]

File: .github/workflows/js_tests.yml (hunk 5 above)
Comment (Ron-Lavi) [reply .github/workflows/js_tests.yml]: makes sense :)

**Review by ekohl — DISMISSED** [review]
From a CI perspective :+1: but perhaps @sharvit or @amirfefer can take a look as well since they're more likely to actually use this.

File: .dockerignore [hunk 6]
```diff
@@ -54,6 +54,5 @@ out*
 .byebug_history
 .gitignore
 .dockerignore
-.travis.yml
```
Comment (tbrisker) [inline .dockerignore:57]: Can you add `.github/` instead? seems to be missing here

Comment (Ron-Lavi) [reply .dockerignore:57]: sure, done :+1:

**Review by ekohl — DISMISSED** [review]
Good catch @tbrisker.

File: .github/workflows/js_tests.yml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+        with:
+          node-version: ${{ matrix.node-version }}
+      - name: Install npm dependencies
+        run: npm install
+      - name: Run linter
+        run: npm run lint
+      - name: Run tests
+        run: npm run test
+      - name: Publish Coveralls
+        uses: coverallsapp/github-action@master
```
Comment (sharvit) [inline .github/workflows/js_tests.yml:33]: Coveralls have a special configuration when using `matrix`, see: https://github.com/marketplace/actions/coveralls-github-action#complete-parallel-job-example

File: .github/workflows/js_tests.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        node-version:
+          - 10
+          - 12
```
Comment (sharvit) [inline .github/workflows/js_tests.yml:16]: Would be nice to add `14` as well

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @laviro 👍 Let's drop travis

File: .github/workflows/js_tests.yml (hunk 8 above)
Comment (Ron-Lavi) [reply .github/workflows/js_tests.yml:16]: is it out already? :open_mouth:  added :+1:

File: .github/workflows/js_tests.yml (hunk 7 above)
Comment (Ron-Lavi) [reply .github/workflows/js_tests.yml:33]: good catch, added
lets see if it works as expected

File: webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/__snapshots__/BookmarkForm.test.js.snap [hunk 9]
```diff
@@ -67,7 +67,6 @@ exports[`BookmarkForm should render bookmarks form with initial values 1`] = `
           "transforms": Array [
             [Function],
           ],
-          "type": "string",
```
Comment (Ron-Lavi) [inline webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/__snapshots__/BookmarkForm.test.js.snap]: Fixed failed snapshot - not related to this PR

Comment (tbrisker) [reply webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/__snapshots__/BookmarkForm.test.js.snap]: the fix was just merged, please rebase on develop.

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/__snapshots__/BookmarkForm.test.js.snap]: rebased, and I still see it, also locally

Comment (tbrisker) [reply webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/__snapshots__/BookmarkForm.test.js.snap]: @MariaAga can you take a look please?

Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/__snapshots__/BookmarkForm.test.js.snap]: Is the vendor version is `4.14.6`? might need to remove `node_modules` and `npm i` to update

Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/components/Bookmarks/components/BookmarkForm/__tests__/__snapshots__/BookmarkForm.test.js.snap]: thanks, before today's last forced push I removed node_modules,
but there seem last run was just between the new foreman-vendor release and Maria fix,
now it works locally, lets wait for tests here.

**Review by sharvit — DISMISSED** [review]
Thanks @laviro LGTM 👍

**Comment by tbrisker** [comment]
Looks like we get 4 coverall statuses now, can we reduce that to just one? also, would it be possible to have just one status for the js tests instead of 4?

**Comment by Ron-Lavi** [comment]
@tbrisker I am not sure how to reduce the statuses amount,
I think this is how the GH actions works with very detailed status per test,
The coveralls part was extended when running them in parallel per node version

**Comment by ekohl** [comment]
Do we need coveralls? I always think the service is just very annoying. The coverage percentage says IMHO absolutely nothing. You can easily get 100% coverage and still miss all the important bugs. I'd like to get rid of coveralls.

**Comment by Ron-Lavi** [comment]
@ekohl IMO it helps to detect dead code or missing unit tests which is good to catch many small issues that while doing refactores can catch many bugs. But maybe we should discuss it with a broader audience.

**Comment by Ron-Lavi** [comment]
Maybe we should drop node 10 support? As earlier we supported only the 2 latest stable versions?
It will also remove 1 test and 2 statuses..

**Comment by ekohl** [comment]
I think our packaging (both deb and RPM) use Node 12 nowadays so it's probably safe to drop 10.

**Comment by Ron-Lavi** [comment]
Rebased and dropped node 10 from tests.

**Comment by ezr-ondrej** [comment]
I really like this, great job @laviro!
But, the output is not ideal. We have 6 results from that 3 are completely useless :thinking: 
Can we somehow not report from the first two coverall runs ( I think that what it's supposed to do ) and not report status of the `finish` job as it should report just the coveralls. If we can down it to 3 outputs, I'm ok with that, even though it'd be great to have just 2 (coveralls and Jest)

EDIT: Maybe just get rid of coveralls https://github.com/theforeman/foreman/pull/7929#issuecomment-688342107 ? It's cool badge, but that's really all it is (maybe sometimes poking the developer to write tests if (s)he didn't but that's not worth three check results)

EDIT2: if we can't (and from my quick research it's impasible to have job without output as a Check) I'd be inclined to use just one node version for now.

**Comment by Ron-Lavi** [comment]
I think we should remove the coveralls in parallel checks,
this will remove the 2 added coveralls statuses.

**Comment by ezr-ondrej** [comment]
> I think we should remove the coveralls in parallel checks,
> this will remove the 2 added coveralls statuses.

And will the coverall work? Because it won't have anything to report, right?
Maybe there is a way to disable the Check Run reports from the parrallel runs for it?

But IMHO testing against one NodeJS version, given we run on that version is fine.

**Comment by Ron-Lavi** [comment]
Trying to run coveralls only in the test with node v14

**Comment by Ron-Lavi** [comment]
and now we have only 1 coveralls status instead of 3 :+1:

**Review by tbrisker — APPROVED** [review]
Great job @laviro :1st_place_medal: !

**Comment by Ron-Lavi** [comment]
Thanks @tbrisker , now I will rebase spellcheck PR #7989

**Pull request merged**